storage/remote/write_hander.go

```
@@ -0,0 +1,87 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage"
+)
+
+type handler struct {
+	logger     log.Logger
+	appendable storage.Appendable
+}
+
+// NewWriteHandler creates a http.Handler that accepts remote write requests and
+// writes them to the provided appendable.
+func NewWriteHandler(logger log.Logger, appendable storage.Appendable) http.Handler {
+	return &handler{
+		logger:     logger,
+		appendable: appendable,
+	}
+}
+
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	req, err := DecodeWriteRequest(r.Body)
+	if err != nil {
+		level.Error(h.logger).Log("msg", "Error decoding remote write request", "err", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	err = h.write(r.Context(), req)
+	switch err {
+	case nil:
+	case storage.ErrOutOfOrderSample, storage.ErrOutOfBounds, storage.ErrDuplicateSampleForTimestamp:
+		// Indicated an out of order sample is a bad request to prevent retires.
+		level.Error(h.logger).Log("msg", "Out of order sample from remote write", "err", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	default:
+		level.Error(h.logger).Log("msg", "Error appending remote write", "err", err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func (h *handler) write(ctx context.Context, req *prompb.WriteRequest) error {
+	var err error
+	app := h.appendable.Appender(ctx)
+	defer func() {
+		if err != nil {
+			app.Rollback()
+			return
+		}
+		err = app.Commit()
```

don't we want to return an error if `Commit()` failed?

---

err is the named return so it would do that

---

hmm I don't see that `err` is a named return value in the function declaration. Am I missing something obvious?

---

indeed err should be a named return

---

sorry, I have misread

---

Good catch!